src/workerd/io/io-context.h

```diff
@@ -491,6 +491,14 @@ class IoContext final: public kj::Refcounted, private kj::TaskSet::ErrorHandler
   // Arrange for the given promise to execute as part of this request. It will be canceled if the
   // request is canceled.
 
+  void runTask(auto fn, auto&&...args) {
```

Comment:
This is a bit of a weird abstraction. It seems to be giving me two ways to capture values: I can pass in a lambda with captures, or I can pass arguments that in turn get passed as args to the function. Of course if I do pass a lambda with captures, and the lambda is a coroutine, I'm going to run into the coroutine-capture problem (the captures become UAF after the first `co_await`). But the capture-by-args approach feels very unusual and verbose.

Can we perhaps take a function only, and automatically wrap it in `kj::coCapture()` to solve the coroutine-capture problem?

The automatically filling in the first parameter with `IoContext&` also feels weird -- why wouldn't I just capture it? I guess the justification is that call sites might commonly be using `IoContext::current().addTask()` and so factoring out a local variable first is cumbersome? But if they already have said local variable, or they don't need the IoContext inside the callback body, then forcing them to declare the argument is cumbersome.

Reply:
>  It seems to be giving me two ways to capture values: I can pass in a lambda with captures, or I can pass arguments that in turn get passed as args to the function. Of course if I do pass a lambda with captures, and the lambda is a coroutine, I'm going to run into the coroutine-capture problem (the captures become UAF after the first co_await). But the capture-by-args approach feels very unusual and verbose.

We can definitely iterate on this further but the reason I went with this approach is that it doesn't require a lambda at all, as with the example of using `updateStorageWriteUnit`. A lambda *can* be used but it doesn't need to.

That said, I would consider this particular abstraction optional here and don't mind backing it out. It's unlikely to be used that extensively and may not, on it's own, provide enough value for just a couple handful of calls.

Reply:
Right, I understand it can be a plain function instead of a lambda, but I feel like that actually goes against the grain of modern C++. Lambdas were designed to be a better alternative to callbacks. Here I feel like you're sort of reinventing lambdas by capturing through variatic arguments, but it's a lot more verbose that way compared to lambda captures.

I don't feel like this is the right way to be solving the lambda capture problem, I'd prefer to use kj::coCapture.

Reply:
heh, ok. I would tend to have the exact opposite view. I find the capturing lambda's and coCapture to be far less readable lol. That said, it's not worth holding up this change for this so I'll back this particular bit out.